internal/util/function/function_executor.go

```diff
@@ -0,0 +1,257 @@
+/*
+ * # Licensed to the LF AI & Data foundation under one
+ * # or more contributor license agreements. See the NOTICE file
+ * # distributed with this work for additional information
+ * # regarding copyright ownership. The ASF licenses this file
+ * # to you under the Apache License, Version 2.0 (the
+ * # "License"); you may not use this file except in compliance
+ * # with the License. You may obtain a copy of the License at
+ * #
+ * #     http://www.apache.org/licenses/LICENSE-2.0
+ * #
+ * # Unless required by applicable law or agreed to in writing, software
+ * # distributed under the License is distributed on an "AS IS" BASIS,
+ * # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * # See the License for the specific language governing permissions and
+ * # limitations under the License.
+ */
+
+package function
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/pkg/mq/msgstream"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+)
+
+type Runner interface {
+	GetSchema() *schemapb.FunctionSchema
+	GetOutputFields() []*schemapb.FieldSchema
+
+	MaxBatch() int
+	ProcessInsert(inputs []*schemapb.FieldData) ([]*schemapb.FieldData, error)
+	ProcessSearch(placeholderGroup *commonpb.PlaceholderGroup) (*commonpb.PlaceholderGroup, error)
+	ProcessBulkInsert(inputs []storage.FieldData) (map[storage.FieldID]storage.FieldData, error)
+}
+
+type FunctionExecutor struct {
+	runners map[int64]Runner
+}
+
+func createFunction(coll *schemapb.CollectionSchema, schema *schemapb.FunctionSchema) (Runner, error) {
+	switch schema.GetType() {
+	case schemapb.FunctionType_BM25: // ignore bm25 function
+		return nil, nil
+	case schemapb.FunctionType_TextEmbedding:
+		f, err := NewTextEmbeddingFunction(coll, schema)
+		if err != nil {
+			return nil, err
+		}
+		return f, nil
+	default:
+		return nil, fmt.Errorf("unknown functionRunner type %s", schema.GetType().String())
+	}
+}
+
+func CheckFunctions(schema *schemapb.CollectionSchema) error {
+	for _, fSchema := range schema.Functions {
+		if _, err := createFunction(schema, fSchema); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewFunctionExecutor(schema *schemapb.CollectionSchema) (*FunctionExecutor, error) {
+	// If the function's outputs exists in outputIDs, then create the function
```

Comment:
comment is incorrect? this function doesn't have argument `outputIDs`

Reply:
The code should have changed, but the comments have not changed